addons/mail/models/mail_channel.py

```diff
@@ -322,40 +322,48 @@ def add_members(self, partner_ids=None, guest_ids=None, invite_to_rtc_call=False
-                notification = _('<div class="o_mail_notification">joined the channel</div>')
-            else:
-                notification = _(
-                    '<div class="o_mail_notification">invited <a href="#" data-oe-model="res.partner" data-oe-id="%(new_partner_id)d">%(new_partner_name)s</a> to the channel</div>',
-                    new_partner_id=channel_partner.partner_id.id,
-                    new_partner_name=channel_partner.partner_id.name,
-                )
-            channel_partner.channel_id.message_post(body=notification, message_type="notification", subtype_xmlid="mail.mt_comment", notify_by_email=False)
-            new_partner_data = {
-                'id': channel_partner.partner_id.id,
-                'im_status': channel_partner.partner_id.im_status,
-                'name': channel_partner.partner_id.name,
+            channel_data = {
+                'id': channel.id,
+                'guestMembers': [('insert', [])],
+                'memberCount': channel.member_count,
+                'members': [('insert', [])],
             }
-            self.env['bus.bus'].sendone((self._cr.dbname, 'mail.channel', channel_partner.channel_id.id), {
+            new_members = self.env['mail.channel.partner'].sudo().create(members_to_create)
+            for channel_partner in new_members.filtered(lambda channel_partner: channel_partner.partner_id):
+                user = channel_partner.partner_id.user_ids[0] if channel_partner.partner_id.user_ids else self.env['res.users']
+                # notify invited members through the bus
+                if user:
+                    self.env['bus.bus'].sendone((self._cr.dbname, 'res.partner', channel_partner.partner_id.id), {
+                        'type': 'mail.channel_joined',
+                        'payload': {
+                            'channel': channel_partner.channel_id.with_user(user).with_context(allowed_company_ids=user.company_ids.ids).sudo().channel_info()[0],
+                            'invited_by_user_id': self.env.user.id,
+                        },
+                    })
+                # notify existing members with a new message in the channel
+                if channel_partner.partner_id == self.env.user.partner_id:
+                    notification = _('<div class="o_mail_notification">joined the channel</div>')
+                else:
+                    notification = _(
+                        '<div class="o_mail_notification">invited <a href="#" data-oe-model="res.partner" data-oe-id="%(new_partner_id)d">%(new_partner_name)s</a> to the channel</div>',
+                        new_partner_id=channel_partner.partner_id.id,
+                        new_partner_name=channel_partner.partner_id.name,
+                    )
+                channel_partner.channel_id.message_post(body=notification, message_type="notification", subtype_xmlid="mail.mt_comment", notify_by_email=False)
+                channel_data['members'][0][1].append({
```

Comment:
This format is too complex to maintain, create the list at the top instead, write directly on it, and write the list on the data/command when it is done, at the end.